(GH-2933) Create "Debugging plans" documentation (#2970)
puppetlabs/bolt: documentation/debugging_plans.md

# Debugging plans

By default, Bolt does not print the result for each step of a plan to standard
output (stdout). However, you can use one of the following methods to

> investigate a plan's execution:

Author: Ooo, good to know!

- You can view the default debug log, which Bolt prints to each time you run a
  Bolt command.

- You can adjust your log level for detailed information on how Bolt is
  executing your plan, including the results returned from each step.

- You can print logs using built-in plan functions.

- You can print verbose output, which includes the results of each step in a
  plan, by running the plan in verbose mode.

- You can print any value or the result of any step to stdout using built-in
  plan functions.

## Logs

### Default debug log

Bolt logs additional information about a plan run, including output sent to
standard error (stderr), at the `debug` level. When you run a Bolt command, Bolt

> automatically prints to a debug log located in the root of your project named
> `bolt-debug.log`. This log includes all messages printed at the `debug` level
> or higher.

Reviewer: ```suggestion
automatically prints to a debug log named `bolt-debug.log`, 
located in the root of your project. This log includes all messages printed
at the `debug` level or higher.
```

### Setting the log level

You can adjust the log level that Bolt prints at using the `log-level`
command-line option:

_\*nix shell command_

```shell
$ bolt plan run example --targets example.org --log-level trace
```

_PowerShell cmdlet_

```powershell
> Invoke-BoltPlan -Name example -Targets example.org -LogLevel trace
```

### Default action logging

Bolt logs actions that a plan takes on targets, such as running commands,
scripts, or tasks, or downloading or uploading files. By default, Bolt logs
an `info` level message when an action starts and another when it completes. If

> you pass a description to the action, that is used in place of the generic log
> message.

Reviewer: ```suggestion
you pass a description to the action, Bolt uses the description in place of 
the generic log message.
```

_Puppet language plan_

```puppet
run_task('my_task', $targets, 'Better description', 'param1' => 'val')
```

_YAML plan_

```yaml
steps:
  - description: Better description
    task: my_task
    targets: $targets
    parameters:
      param1: val
```

If your plan contains many small actions, you might want to suppress these

> messages and use explicit calls to log functions instead. This can be
> accomplished by wrapping actions in a `without_default_logging` block in Puppet

Reviewer: ```suggestion
messages and use explicit calls to log functions instead. You can accomplish
this by wrapping actions in a `without_default_logging` block in Puppet
```

# Debugging plans

By default, Bolt does not print the result for each step of a plan to standard
output (stdout). However, there are multiple ways to debug a Bolt plan.

> You can investigate a plan's execution by:

Reviewer: ```suggestion
To investigate a plan's execution:
```
My bad 😅
